primitives/runtime/src/vec_db.rs

```diff
@@ -0,0 +1,332 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Logic for working with more efficient storage proofs.
+
+use frame_support::{PalletError, StateVersion};
+use sp_core::{storage::TrackedStorageKey, RuntimeDebug};
+use sp_runtime::SaturatedConversion;
+use sp_std::{default::Default, vec, vec::Vec};
+use sp_trie::{
+	generate_trie_proof, verify_trie_proof, LayoutV0, LayoutV1, StorageProof, TrieDBBuilder,
+	TrieHash,
+};
+
+use codec::{Decode, Encode};
+use hash_db::Hasher;
+use scale_info::TypeInfo;
+use trie_db::{DBValue, Trie};
+
+use crate::{storage_proof::RawStorageProof, Size};
+
+pub type RawStorageKey = Vec<u8>;
+
+/// Errors that can occur when interacting with `UntrustedVecDb` and `TrustedVecDb`.
+#[derive(Clone, Encode, Decode, RuntimeDebug, PartialEq, Eq, PalletError, TypeInfo)]
+pub enum VecDbError {
+	/// Call to `generate_trie_proof()` failed.
+	UnableToGenerateTrieProof,
+	/// Call to `verify_trie_proof()` failed.
+	InvalidProof,
+	/// The `Vec` entries weren't sorted as expected.
+	UnsortedEntries,
+	/// The provided key wasn't found.
+	UnavailableKey,
+	/// The value associated to the provided key is `None`.
+	EmptyVal,
+	/// Error decoding value associated to a provided key.
+	DecodeError,
+	/// At least one key in the `VecDb` wasn't read.
+	UnusedKey,
+}
+
+/// Structure representing a key-value database stored as a sorted `Vec` of tuples.
+///
+/// The structure also contains a proof of the fact that the key-value tuples are actually present
+/// in the chain storage.
+#[derive(Clone, Default, Decode, Encode, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct UntrustedVecDb {
+	proof: RawStorageProof,
+	db: Vec<(RawStorageKey, Option<DBValue>)>,
+}
+
+impl UntrustedVecDb {
+	/// Creates a new instance of `UntrustedVecDb`.
+	pub fn new<H: Hasher>(
+		read_proof: StorageProof,
+		root: TrieHash<LayoutV1<H>>,
+		mut keys: Vec<impl AsRef<[u8]> + Ord>,
+	) -> Result<Self, VecDbError> {
+		// It's ok to use `LayoutV1` in this function, no matter the actual underlying layout,
+		// because we only perform read operations. When reading `LayoutV0` and `LayoutV1` lead to
+		// the same result.
+		let mem_db = read_proof.into_memory_db();
+		let trie_db = TrieDBBuilder::<LayoutV1<H>>::new(&mem_db, &root).build();
+
+		let trie_proof = generate_trie_proof::<LayoutV1<H>, _, _, _>(&mem_db, root, &keys)
+			.map_err(|_| VecDbError::UnableToGenerateTrieProof)?;
+
+		let mut entries = vec![];
```

Comment:
Couldnt we initialize `Vec::with_capacity(keys.len())` or something like that?

Reply:
Yes. Missed this. Thanks ! Done.